pkg/server/container_update_resources.go

```diff
@@ -20,41 +20,150 @@ import (
-			Quota:  proto.Int64(r.GetCpuQuota()),
-			Period: proto.Uint64(uint64(r.GetCpuPeriod())),
-			Cpus:   r.GetCpusetCpus(),
-			Mems:   r.GetCpusetMems(),
-		},
-		Memory: &runtimespec.LinuxMemory{
-			Limit: proto.Int64(r.GetMemoryLimitInBytes()),
-		},
+func (c *criContainerdService) updateContainerResources(ctx context.Context,
+	cntr containerstore.Container,
+	resources *runtime.LinuxContainerResources,
+	status containerstore.Status) (retErr error) {
+	id := cntr.ID
+	// Do not update the container when there is a removal in progress.
+	if status.Removing {
+		return fmt.Errorf("container %q is in removing state", id)
+	}
+
+	// Update container spec. If the container is not started yet, updating
+	// spec makes sure that the resource limits are correct when start;
+	// if the container is already started, updating spec is still required,
+	// the spec will become our source of truth for resource limits.
+	oldSpec, err := cntr.Container.Get().Spec()
+	if err != nil {
+		return fmt.Errorf("failed to get container spec: %v", err)
+	}
+	newSpec, err := updateOCILinuxResource(oldSpec, resources)
+	if err != nil {
+		return fmt.Errorf("failed to update resource in spec: %v", err)
+	}
+
+	info := cntr.Container.Get().Info()
+	any, err := typeurl.MarshalAny(newSpec)
+	if err != nil {
+		return fmt.Errorf("failed to marshal spec %+v: %v", newSpec, err)
+	}
+	info.Spec = any
+	// TODO(random-liu): Add helper function in containerd to do the update.
+	if _, err := c.client.ContainerService().Update(ctx, info, "spec"); err != nil {
+		return fmt.Errorf("failed to update container spec: %v", err)
+	}
+	defer func() {
+		if retErr != nil {
+			// Reset spec on error.
+			any, err := typeurl.MarshalAny(oldSpec)
+			if err != nil {
+				glog.Errorf("Failed to marshal spec %+v for container %q: %v", oldSpec, id, err)
+				return
+			}
+			info.Spec = any
+			if _, err := c.client.ContainerService().Update(ctx, info, "spec"); err != nil {
+				glog.Errorf("Failed to recover spec %+v for container %q: %v", oldSpec, id, err)
+			}
+		}
+	}()
+	container, err := c.client.LoadContainer(ctx, id)
+	if err != nil {
+		return fmt.Errorf("failed to load container: %v", err)
+	}
+	defer func() {
+		if retErr == nil {
+			// Update container client if no error is returned.
+			// NOTE(random-liu): By updating container client, we'll be able
+			// to get latest OCI spec from it, which includes the up-to-date
+			// container resource limits. This will be useful after the debug
+			// api is introduced.
+			cntr.Container.Set(container)
+		}
+	}()
+
+	// If container is not running, only update spec is enough, new resource
+	// limit will be applied when container start.
+	if status.State() != runtime.ContainerState_CONTAINER_RUNNING {
+		return nil
+	}
+
+	task, err := container.Task(ctx, nil)
+	if err != nil {
+		if errdefs.IsNotFound(err) {
+			// Task exited already.
+			return nil
+		}
+		return fmt.Errorf("failed to get task: %v", err)
```

Comment:
random thought here.. the term "task" may be very confusing to kubernetes users... Let's consider using "container" or "container state" instead of task in error messages we generate.

Reply:
We have many error messages include task now. :p We may want to address them all when we have better name for it. :)